src/lib.rs

```diff
@@ -0,0 +1,116 @@
+use std::error;
+use std::iter;
+
+/// The basic `Model` trait.
+///
+/// It is training-agnostic: a model takes an input and returns an output.
+///
+/// There might be multiple ways to discover the best settings for every
+/// particular algorithm (e.g. training a logistic regressor using
+/// a pseudo-inverse matrix vs using gradient descent).
+/// It doesn't matter: the end result, the model, is a set of parameters.
+/// The way those parameter originated is an orthogonal concept.
+///
+/// In the same way, it has no notion of loss or "correct" predictions.
+/// Those concepts are embedded elsewhere.
+pub trait Model {
+    type Input;
+    type Output;
+    type Error: error::Error;
+
+    fn predict(&self, inputs: &Self::Input) -> Result<Self::Output, Self::Error>;
+}
+
+/// One step closer to the peak.
+///
+/// `Optimizer` is generic over a type `M` implementing the `Model` trait: `M` is used to
+/// constrain what type of inputs and targets are acceptable.
+///
+/// `train` takes an instance of `M` as one of its inputs, `model`: it doesn't matter if `model`
+/// has been through several rounds of training before, or if it just came out of a `Blueprint`
+/// using `initialize` - it's consumed by `train` and a new model is returned.
+///
+/// This means that there is no difference between one-shot training and incremental training.
+/// Furthermore, the optimizer doesn't have to "own" the model or know anything about its hyperparameters,
+/// because it never has to initialize it.
+pub trait Optimizer<M>
+where
+    M: Model,
+{
+    type Error: error::Error;
+
+    fn train(
+        &self,
+        inputs: &M::Input,
+        targets: &M::Output,
+        model: M,
+    ) -> Result<M, Self::Error>;
+}
+
+/// Where `Model`s are forged.
+///
+/// `Blueprint`s are used to specify how to build and initialize an instance of the model type `M`.
+///
+/// For the same model type `M`, nothing prevents a user from providing more than one `Blueprint`:
+/// multiple initialization strategies can somethings be used to be build the same model type.
+///
+/// Each of these strategies can take different (hyper)parameters, even though they return an
+/// instance of the same model type in the end.
+///
+/// The initialization procedure could be data-dependent, hence the signature of `initialize`.
```

Comment:
I'm a bit concerned about potential user confusion about what should be put in the `Blueprint`'s `initialize` method vs `Optimizer`'s `train` method, given the similarities in method signatures (they both take input and targets, they both return models). 

What would be an example of a workflow with a data-dependent initialization? Are there any other options for handling that initialization?